### PR TITLE
Add in normalising flows functionality as used in Storm's paper

### PR DESCRIPTION
**Pull request by scolloms**
This updates the functionality of the inference to allow running `model_select` with normalising flows, in addition to KDEs. 

These changes hard-code the use of particular population synthesis models from the One Channel paper, and need to be updated to allow for flexible population model input. This concerns `model_select`, `populations/Pop_flows.py`, `populations/utils/flows.py`, `populations/bbh_models.py`, and `sample/sample.py`.

Other changes include adding regularisation to the KDE models in `populations/__init__.py`, and changing the way GW posterior samples are loaded when we require more samples than exists for an event. I also include the updated README, gitignore, environment file, and GW event pre-processing notebook.

### CODE REVIEW

File: model_select [hunk 1]
```diff
@@ -21,21 +21,22 @@ import scipy.stats
...
 from populations import bbh_models, gw_obs
 from sample import sample
-from plot import msplot
 
 # --- Argument handling --- #
 argp = argparse.ArgumentParser()
 argp.add_argument("--file-path", type=str, required=True, help="Path to where the population models are living. Models should be stored in hdf5 format, with the channel as the base group, and model parameters as subgroups (e.g., 'CE/chi01/alpha0.5').")
 argp.add_argument("--model0", type=str, required=True, help="Sets the 'true' model from which mock observations are drawn. When there are multiple parameters in the submodel, separate with '/' (e.g., 'chi00/alpha5'). If 'gwobs', will use the actual GW observations.")
+argp.add_argument("--use-flows", action="store_true", help="True if using normalising flows to estimate hyperlikelihoods in inference. Otherwise use KDEs. Default=False.")
+argp.add_argument("--discrete-sample", action="store_true", help="If true then sample on hyperparameter indices to just evaluate likelihood at locations of training models. Set to True is useflows is False. Default=False.")
```
Comment (michaelzevin) [inline model_select]: Just to clarify, the reason that this is an argument is so that it can be set to TRUE when you're using the flows so that you can still do the discrete inference? 

Since this needs to be TRUE if `use-flows==FALSE`, either this argument or --use-flows should default to True so that the code can be run with default settings.

File: model_select [hunk 2]
```diff
@@ -44,10 +45,26 @@ argp.add_argument("--multiproc", action="store_true", help="Determines whether t
...
+argp.add_argument("--spline-bins", nargs="+", default=[4,4,4,4,4], help="Number of bins to use in Coupling Spline flows. List, must be same length as channels. Default=4.")
+argp.add_argument("--no-trans", nargs="+", default=[6,6,6,6,6], help="Number of transformations to train flows. List, must be same length as channels. Default=6.")
+argp.add_argument("--no-neurons", nargs="+", default=[128,128,128,128,128,128], help="Number of neurons to train flows. List, must be same length as channels. Default=128.")
+argp.add_argument("--batch-no", type=int, default='10', help="Number of batches to train flow models. Default=10.")
+argp.add_argument("--device", type=str, default='cpu', help="Device to run model select on. Default=cpu, otherwise choose cuda:X where X is the slot of the GPU.")
+argp.add_argument("--use-wandb", action="store_true", help="Whether or not to use Weights and biases to run a sweep over optimising parameters of the normalising flows.")
+argp.add_argument("--testCEmodel", action="store_true", help="If true removes the chi_b=0.1, alphaCE=1.0 training samples during the flow training to test interpolation.")
```
Comment (michaelzevin) [inline model_select:60]: We can keep this for now, but probably won't want this functionality in the ini file.

File: model_select [hunk 3]
```diff
@@ -261,44 +318,37 @@ else:
     # assume equal prior weight for each posterior sample
     p_theta = np.ones((obsdata.shape[0],obsdata.shape[1]))
 
-# plot for paper (uncomment if generating)
```
Comment (michaelzevin) [inline model_select]: Can you keep this block in for now? You can set it so that it only runs when flows are turned off. I'll change it up when I do the full code refresh.

File: model_select [hunk 4]
```diff
@@ -307,37 +357,46 @@ while all_models_at_deepest==False:
...
+if args.random_seed:
+    np.random.seed(args.random_seed)
 
 # --- Do the sampling!
-
-sampler = sample.Sampler(model_names)
-sampler.sample(kde_models, obsdata, verbose=verbose)
+#Here we pass different args is we are doing discrete or continuous sampling of the population model parameters
+if discrete_sample:
+    sampler = sample.Sampler(model_names, hyperparam_bounds=[[0, len(hyperparam_dict[0])], [0, len(hyperparam_dict[1])]], likelihood='emcee_lnlike_disc')
```
Comment (michaelzevin) [inline model_select]: This seems like it hardcodes in the number of hyperparameters, right?

File: model_select [hunk 5]
```diff
@@ -347,44 +406,6 @@ raw_samples = raw_samples.reshape((raw_samples.shape[0] * raw_samples.shape[1],
 lnprb = lnprb.reshape((lnprb.shape[0] * lnprb.shape[1]))
 
 
-# Get samples for both *detectable* branching fractions and *underlying* branching fractions
```
Comment (michaelzevin) [inline model_select:350]: Can you also keep this block of code in for now? You can put it behind a boolean for the KDE discrete sampling.

File: populations/bbh_models.py [hunk 6]
```diff
@@ -44,9 +44,76 @@ def get_params(df, params):
...
+    f.close()
 
+    # use only models with given alpha value
+    for model in all_models:
+        if channel in model:
+            models.append('/'+model)
+    return(np.array(models))
 
-def get_models(file_path, channels, params, spin_distr=None, sensitivity=None, normalize=False, detectable=False, **kwargs):
+def get_model_keys_CE(path):
```
Comment (michaelzevin) [inline populations/bbh_models.py:67]: It would be good to generalize this so we don't hardcode the CE channel as such, and instead have the code understand how deep each model is. But I'll merge for now and then we can revisit this.

**Review by michaelzevin — CHANGES_REQUESTED** [review]

File: model_select (hunk 1 above)
Comment (scolloms) [reply model_select]: Resolved by switching from store true arg `discrete-sample` to `continuous-sample` in commit https://github.com/michaelzevin/model_selection/pull/9/commits/f48029103d53488895eefe1450f8f477733fb759

File: model_select (hunk 2 above)
Comment (scolloms) [reply model_select:60]: `batch-no` should be kept as an argument (and should be added to the flow architecture config file set-up). `device` I think needs to be kept for using pytorch with GPUs, though I imagine the ini file should include a `require_gpus` argument. `use-wandb` and `testCEmodel` should be removed for the main release, though if we want to release something where the user can optimise the normlaiisng flow hyperparameters, we can provide a guide/seperate release for how to do this?

File: model_select (hunk 3 above)
Comment (scolloms) [reply model_select]: Done in commit https://github.com/michaelzevin/model_selection/pull/9/commits/f48029103d53488895eefe1450f8f477733fb759

File: model_select (hunk 4 above)
Comment (scolloms) [reply model_select]: Yes, I have updated this in commit https://github.com/scolloms/AMAZE_model_selection/commit/d7cc79267902f14bf158cf4eb617c1eeb5bc7ef0 to allow for Nhyper number of hyperparameters. However priors in `sample.py` still only handle Nhyper=2, so this needs changing

File: model_select (hunk 5 above)
Comment (scolloms) [reply model_select:350]: Done in commit https://github.com/michaelzevin/model_selection/commit/f48029103d53488895eefe1450f8f477733fb759

File: populations/bbh_models.py (hunk 6 above)
Comment (scolloms) [reply populations/bbh_models.py:67]: Yes, intending on changing this alongside other changes to remove hardcoding of channels in the flow scripts

**Comment by scolloms** [comment]
Addressed @michaelzevin's comments in commits https://github.com/michaelzevin/model_selection/pull/9/commits/f48029103d53488895eefe1450f8f477733fb759 and https://github.com/michaelzevin/model_selection/pull/9/commits/d7cc79267902f14bf158cf4eb617c1eeb5bc7ef0, this should now be ready to merge

**Review by michaelzevin — APPROVED** [review]
